From 9585e38b2a533422ee76a414c4ac38d8e59953d7 Mon Sep 17 00:00:00 2001
From: Michael Bestas <mkbestas@gmail.com>
Date: Mon, 5 Oct 2026 16:53:15 +0300
Subject: [PATCH] AppWidgetServiceImpl: Fix disappearing widgets on devices
 with a work profile

This matches Pixel CP2A.260705.006 behavior:
* stopListening, setAppWidgetHidden: bail out for locked users
* queryIntentReceivers: unconditionally match direct-boot aware/unaware
  receivers and log the result

Reference: https://www.androidenterprise.community/android-enterprise-general-discussions-3/widgets-missing-after-android-17-upgrade-on-devices-with-a-work-profile-2720
Change-Id: I48e10125548e9826b1c4b2f246616c0a15808822
---
 .../appwidget/AppWidgetServiceImpl.java       | 37 +++++++++----------
 1 file changed, 18 insertions(+), 19 deletions(-)

diff --git a/services/appwidget/java/com/android/server/appwidget/AppWidgetServiceImpl.java b/services/appwidget/java/com/android/server/appwidget/AppWidgetServiceImpl.java
index 15f4aa4f1833..09fec3ae7479 100644
--- a/services/appwidget/java/com/android/server/appwidget/AppWidgetServiceImpl.java
+++ b/services/appwidget/java/com/android/server/appwidget/AppWidgetServiceImpl.java
@@ -1453,6 +1453,11 @@ class AppWidgetServiceImpl extends IAppWidgetService.Stub implements WidgetBacku
         // Make sure the package runs under the caller uid.
         mSecurityPolicy.enforceCallFromPackage(callingPackage);
 
+        if (!isUserRunningAndUnlocked(userId)) {
+            Slog.w(TAG, "stopListening() should not be called for locked user " + userId);
+            return;
+        }
+
         synchronized (mLock) {
             ensureGroupStateLoadedLocked(userId, /* enforceUserUnlockingOrUnlocked */ false);
 
@@ -1535,6 +1540,11 @@ class AppWidgetServiceImpl extends IAppWidgetService.Stub implements WidgetBacku
 
         mSecurityPolicy.enforceCallFromPackage(callingPackage);
 
+        if (!isUserRunningAndUnlocked(userId)) {
+            Slog.w(TAG, "setAppWidgetHidden() should not be called for locked user " + userId);
+            return;
+        }
+
         synchronized (mLock) {
             ensureGroupStateLoadedLocked(userId, /* enforceUserUnlockingOrUnlocked */false);
 
@@ -4191,21 +4201,20 @@ class AppWidgetServiceImpl extends IAppWidgetService.Stub implements WidgetBacku
             // provide widgets.
             flags |= PackageManager.MATCH_DEBUG_TRIAGED_MISSING;
 
-            // Widget hosts that are non-crypto aware may be hosting widgets
-            // from a profile that is still locked, so let them see those
-            // widgets.
-            if (isProfileWithUnlockedParent(userId)) {
-                flags |= PackageManager.MATCH_DIRECT_BOOT_AWARE
-                        | PackageManager.MATCH_DIRECT_BOOT_UNAWARE;
-            }
+            // Match receivers regardless of the user's lock state.
+            flags |= PackageManager.MATCH_DIRECT_BOOT_AWARE
+                    | PackageManager.MATCH_DIRECT_BOOT_UNAWARE;
 
             // Widgets referencing shared libraries need to have their
             // dependencies loaded.
             flags |= PackageManager.GET_SHARED_LIBRARY_FILES;
 
-            return mPackageManager.queryIntentReceivers(intent,
+            List<ResolveInfo> receivers = mPackageManager.queryIntentReceivers(intent,
                     intent.resolveTypeIfNeeded(mContext.getContentResolver()),
                     flags, userId).getList();
+            Slog.i(TAG, "queryIntentReceivers for user " + userId + " found: "
+                    + receivers.size() + " receivers.");
+            return receivers;
         } catch (RemoteException re) {
             Slog.w(TAG, "Failed to query intent receivers for user " + userId, re);
             return null;
@@ -4238,6 +4247,7 @@ class AppWidgetServiceImpl extends IAppWidgetService.Stub implements WidgetBacku
             Trace.traceEnd(Trace.TRACE_TAG_ACTIVITY_MANAGER);
 
             final int N = mProviders.size();
+            Slog.i(TAG, "handleUserUnlocked: " + userId + " mProviders size: " + N);
             for (int i = 0; i < N; i++) {
                 Provider provider = mProviders.get(i);
 
@@ -5262,17 +5272,6 @@ class AppWidgetServiceImpl extends IAppWidgetService.Stub implements WidgetBacku
         return false;
     }
 
-    private boolean isProfileWithUnlockedParent(int userId) {
-        UserInfo userInfo = mUserManager.getUserInfo(userId);
-        if (userInfo != null && userInfo.isProfile()) {
-            UserInfo parentInfo = mUserManager.getProfileParent(userId);
-            if (parentInfo != null
-                    && mUserManager.isUserUnlockingOrUnlocked(parentInfo.getUserHandle())) {
-                return true;
-            }
-        }
-        return false;
-    }
 
     /**
      * Note an app widget is tapped on. If a app widget is tapped, the underlying app is treated as
-- 
2.56.0

